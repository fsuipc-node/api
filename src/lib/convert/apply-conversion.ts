import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
