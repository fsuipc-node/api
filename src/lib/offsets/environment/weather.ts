import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const weather: OffsetList = {
  metarStationAltitude: new Offset({
    value: 0x4B4,
    name: 'metarStationAltitude',
    category: OffsetCategory.WEATHER,
    description: 'METAR station altitude - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  windSurfaceTurbulenceGustSpeed: new Offset({
    value: 0x4BA,
    name: 'windSurfaceTurbulenceGustSpeed',
    category: OffsetCategory.WEATHER,
    description: 'METAR station altitude - kt',
    type: Type.UInt16,
    convert: '+({VAL}).toFixed(2)',
    permission: 'r',
  }),
  metarBarometricDrift: new Offset({
    value: 0x4BC,
    name: 'metarBarometricDrift',
    category: OffsetCategory.WEATHER,
    description: 'METAR barometric drift - difference between aircraft/METAR QNH. Adding drift will give correct value for ATIS report',
    type: Type.UInt16,
    convert: '{VAL} / 16',
    permission: 'r',
  }),
  metarVisibility: new Offset({
    value: 0x4C0,
    name: 'metarVisibility',
    category: OffsetCategory.WEATHER,
    description: 'METAR visibility - sm',
    type: Type.UInt16,
    convert: '{VAL} / 100',
    permission: 'r',
  }),
  metarCloudThunderBase: new Offset({
    value: 0x4C2,
    name: 'metarCloudThunderBase',
    category: OffsetCategory.WEATHER,
    description: 'METAR visibility - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  metarCloudBaseLow: new Offset({
    value: 0x4C4,
    name: 'metarCloudBaseLow',
    category: OffsetCategory.WEATHER,
    description: 'METAR cloud base low - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  metarCloudBaseHigh: new Offset({
    value: 0x4C6,
    name: 'metarCloudBaseHigh',
    category: OffsetCategory.WEATHER,
    description: 'METAR cloud base high - ft',
    type: Type.UInt16,
    convert: '+({VAL} * 3.28084).toFixed(2)',
    permission: 'r',
  }),
  dewPoint: new Offset({
    value: 0x4C8,
    name: 'dewPoint',
    category: OffsetCategory.WEATHER,
    description: 'dew point - FS2000+',
    type: Type.Int16,
    convert: '{VAL} / 256',
    permission: 'r',
  }),
  precipitationRate: new Offset({
    value: 0x4C8,
    name: 'precipitationRate',
    category: OffsetCategory.WEATHER,
    description: 'precipitation rate - FS2000+ - 0 to 5',
    type: Type.Byte,
    permission: 'r',
  }),
  precipitationType: new Offset({
    value: 0x4CC,
    name: 'precipitationType',
    category: OffsetCategory.WEATHER,
    description: 'precipitation type - FS2000+ - 0 to 5',
    type: Type.Byte,
    convert: 'precipitationType',
    mapping: true,
    permission: 'r',
  }),
  cloudThunderCoverage: new Offset({
    value: 0x4CD,
    name: 'cloudThunderCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud thunder coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudLowCoverage: new Offset({
    value: 0x4CE,
    name: 'cloudLowCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud low coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudHighCoverage: new Offset({
    value: 0x4CF,
    name: 'cloudHighCoverage',
    category: OffsetCategory.WEATHER,
    description: 'cloud high coverage - octa',
    type: Type.Byte,
    permission: 'r',
  }),
  precipitationControl: new Offset({
    value: 0x4D2,
    name: 'precipitationControl',
    category: OffsetCategory.WEATHER,
    description: 'hi-byte type 0-2 - low-byte rate 0-5 - writing 0xFFFF release control',
    type: Type.UInt16,
    permission: 'w',
  }),
  dewPointControl: new Offset({
    value: 0x4D4,
    name: 'dewPointControl',
    category: OffsetCategory.WEATHER,
    description: 'degrees / 256 - 0x8000 release control',
    type: Type.UInt16,
    permission: 'w',
  }),
  surfaceWindSpeed: new Offset({
    value: 0x4D8,
    name: 'surfaceWindSpeed',
    category: OffsetCategory.WEATHER,
    description: 'surface wind speed - FS2000+ - kts',
    type: Type.UInt16,
    permission: 'r',
  }),
  surfaceWindDirection: new Offset({
    value: 0x4DA,
    name: 'surfaceWindDirection',
    category: OffsetCategory.WEATHER,
    description: 'surface wind direction - FS2000+ - deg MAG',
    type: Type.UInt16,
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    permission: 'r',
  }),
  unlimitedVisibility: new Offset({
    value: 0xBF8,
    name: 'unlimitedVisibility',
    category: OffsetCategory.WEATHER,
    description: 'unlimited visibility value - sm',
    type: Type.Int32,
    convert: '+({VAL} / 1600).toFixed(2)',
    permission: 'rw',
  }),
  cloudTypeAtAircraftAltitude: new Offset({
    value: 0xE84,
    name: 'cloudTypeAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud type at aircraft atltitude - 0-10',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudCoverageAtAircraftAltitude: new Offset({
    value: 0xE85,
    name: 'cloudCoverageAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud coverage at aircraft atltitude - okta',
    type: Type.Byte,
    permission: 'r',
  }),
  cloudIcingAtAircraftAltitude: new Offset({
    value: 0xE86,
    name: 'cloudIcingAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud icing at aircraft atltitude - 0-4',
    type: Type.UInt16,
    permission: 'r',
  }),
  cloudTurbulenceAtAircraftAltitude: new Offset({
    value: 0xE88,
    name: 'cloudTurbulenceAtAircraftAltitude',
    category: OffsetCategory.WEATHER,
    description: 'cloud turbulence at aircraft atltitude - 0-255',
    type: Type.UInt16,
    permission: 'r',
  }),
  visibility: new Offset({
    value: 0xE8A,
    name: 'visibility',
    category: OffsetCategory.WEATHER,
    description: 'current visibility - sm',
    convert: '+({VAL} / 100).toFixed(2)',
    type: Type.Int16,
    permission: 'r',
  }),
  OAT: new Offset({
    value: 0xE8C,
    name: 'OAT',
    category: OffsetCategory.WEATHER,
    description: 'current outside air temperature (OAT) - celsius',
    convert: 'Math.round({VAL} / 256)',
    type: Type.Int16,
    permission: 'r',
  }),
  interpolatedDewPoint: new Offset({
    value: 0xE8E,
    name: 'interpolatedDewPoint',
    category: OffsetCategory.WEATHER,
    description: 'dew point interpolated for aircraft altitude - celsius',
    convert: 'Math.round({VAL} / 256)',
    type: Type.Int16,
    permission: 'r',
  }),
  windSpeed: new Offset({
    value: 0xE90,
    name: 'windSpeed',
    category: OffsetCategory.WEATHER,
    description: 'wind speed at aircraft - kt',
    type: Type.UInt16,
    permission: 'r',
  }),
  windDirection: new Offset({
    value: 0xE92,
    name: 'windDirection',
    category: OffsetCategory.WEATHER,
    description: 'wind direction at aircraft - TRUE degrees',
    convert: '+({VAL} * 360 / 65536).toFixed(2)',
    type: Type.UInt16,
    permission: 'r',
  }),
};
