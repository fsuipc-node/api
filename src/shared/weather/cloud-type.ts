export enum CloudType {
  USER = 'user',
  CIRRUS = 'cirrus',
  STRATUS = 'stratus',
  CUMULUS = 'cumulus',
}
