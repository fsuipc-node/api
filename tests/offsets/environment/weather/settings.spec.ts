import { weatherSettings as offsets } from '@offsets/environment/weather/settings';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - environment/weather/settings', () => {
  const offsetsTestCases = [
    { name: 'upperCloudCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'upperCloudBaseSetting', value: 914, expectedResult: 2998.69 },
    { name: 'upperCloudCoverageSetting', value: 32768 , expectedResult: 4 },
    { name: 'upperCloudVariationSetting', value: 914, expectedResult: 2998.69 },
    { name: 'lowerCloudCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'lowerCloudBaseSetting', value: 914, expectedResult: 2998.69 },
    { name: 'lowerCloudCoverageSetting', value: 32768 , expectedResult: 4 },
    { name: 'lowerCloudVariationSetting', value: 914, expectedResult: 2998.69 },
    { name: 'stormCloudCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'stormCloudBaseSetting', value: 914, expectedResult: 2998.69 },
    { name: 'stormCloudCoverageSetting', value: 32768 , expectedResult: 4 },
    { name: 'stormCloudVariationSetting', value: 914, expectedResult: 2998.69 },
    { name: 'upperTemperatureLevelSetting', value: 914, expectedResult: 2998.69 },
    { name: 'upperTemperatureSetting', value: 512, expectedResult: 2 },
    { name: 'middleTemperatureLevelSetting', value: 914, expectedResult: 2998.69 },
    { name: 'middleTemperatureSetting', value: 512, expectedResult: 2 },
    { name: 'lowerTemperatureLevelSetting', value: 914, expectedResult: 2998.69 },
    { name: 'lowerTemperatureSetting', value: 512, expectedResult: 2 },
    { name: 'surfaceTemperatureLevelSetting', value: 914, expectedResult: 2998.69 },
    { name: 'surfaceTemperatureSetting', value: 512, expectedResult: 2 },
    { name: 'temperatureDriftSetting', value: 512, expectedResult: 2 },
    { name: 'temperatureDayNightVariationSetting', value: 512, expectedResult: 2 },
    { name: 'pressureQNHSetting', value: 32, expectedResult: 2 },
    { name: 'pressureDriftSetting', value: 32, expectedResult: 2 },
    { name: 'upperWindCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'upperWindBaseSetting', value: 914, expectedResult: 2998.69 },
    { name: 'upperWindDirectionSetting', value: 32768, expectedResult: 180 },
    { name: 'upperWindGustEnabledSetting', value: 1, expectedResult: true },
    { name: 'upperWindGustEnabledSetting', value: 0, expectedResult: false },
    { name: 'middleWindCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'middleWindBaseSetting', value: 914, expectedResult: 2998.69 },
    { name: 'middleWindDirectionSetting', value: 32768, expectedResult: 180 },
    { name: 'middleWindGustEnabledSetting', value: 1, expectedResult: true },
    { name: 'middleWindGustEnabledSetting', value: 0, expectedResult: false },
    { name: 'lowerWindCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'lowerWindBaseSetting', value: 914, expectedResult: 2998.69 },
    { name: 'lowerWindDirectionSetting', value: 32768, expectedResult: 180 },
    { name: 'lowerWindGustEnabledSetting', value: 1, expectedResult: true },
    { name: 'lowerWindGustEnabledSetting', value: 0, expectedResult: false },
    { name: 'surfaceWindCeilingSetting', value: 914, expectedResult: 2998.69 },
    { name: 'surfaceWindDirectionSetting', value: 32768, expectedResult: 180 },
    { name: 'surfaceWindGustEnabledSetting', value: 1, expectedResult: true },
    { name: 'surfaceWindGustEnabledSetting', value: 0, expectedResult: false },
    { name: 'upperCloudIcingSetting', value: 1, expectedResult: true },
    { name: 'upperCloudIcingSetting', value: 0, expectedResult: false },
    { name: 'lowerCloudIcingSetting', value: 1, expectedResult: true },
    { name: 'lowerCloudIcingSetting', value: 0, expectedResult: false },
    { name: 'stormCloudIcingSetting', value: 1, expectedResult: true },
    { name: 'stormCloudIcingSetting', value: 0, expectedResult: false },
    { name: 'visibilitySetting', value: 1000, expectedResult: 10 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
