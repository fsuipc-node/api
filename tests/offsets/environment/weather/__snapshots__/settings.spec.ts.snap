// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - environment/weather/settings offsets list should have required properties 1`] = `
Object {
  "lowerCloudBaseSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower cloud layer base - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudBaseSetting",
    "permission": "rw",
    "type": 5,
    "value": 3878,
  },
  "lowerCloudCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower cloud layer ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3876,
  },
  "lowerCloudCoverageSetting": Offset {
    "category": "weather",
    "convert": "Math.round(({VAL} + 1) / 8192)",
    "description": "lower cloud layer coverage - oktas - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudCoverageSetting",
    "permission": "rw",
    "type": 5,
    "value": 3880,
  },
  "lowerCloudIcingSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "lower cloud icing - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudIcingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3970,
  },
  "lowerCloudTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "lower cloud turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3972,
  },
  "lowerCloudTypeSetting": Offset {
    "category": "weather",
    "convert": "cloudType",
    "description": "lower cloud type - setting",
    "length": undefined,
    "mapping": true,
    "name": "lowerCloudTypeSetting",
    "permission": "rw",
    "type": 5,
    "value": 3968,
  },
  "lowerCloudVariationSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower cloud layer variation - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerCloudVariationSetting",
    "permission": "rw",
    "type": 5,
    "value": 3882,
  },
  "lowerTemperatureLevelSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower temperature level - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerTemperatureLevelSetting",
    "permission": "rw",
    "type": 5,
    "value": 3900,
  },
  "lowerTemperatureSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "lower temperature - degrees C - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerTemperatureSetting",
    "permission": "rw",
    "type": 2,
    "value": 3902,
  },
  "lowerWindBaseSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindBaseSetting",
    "permission": "rw",
    "type": 5,
    "value": 3942,
  },
  "lowerWindCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "lower wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3940,
  },
  "lowerWindDirectionSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "lower wind ceiling - TRUE degrees - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindDirectionSetting",
    "permission": "rw",
    "type": 5,
    "value": 3946,
  },
  "lowerWindGustEnabledSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "lower wind gust enabled - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindGustEnabledSetting",
    "permission": "rw",
    "type": 2,
    "value": 3950,
  },
  "lowerWindSpeedSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "lower wind ceiling - kt - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindSpeedSetting",
    "permission": "rw",
    "type": 5,
    "value": 3944,
  },
  "lowerWindTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "lower wind turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "lowerWindTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3948,
  },
  "middleTemperatureLevelSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "middle temperature level - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleTemperatureLevelSetting",
    "permission": "rw",
    "type": 5,
    "value": 3896,
  },
  "middleTemperatureSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "middle temperature - degrees C - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleTemperatureSetting",
    "permission": "rw",
    "type": 2,
    "value": 3898,
  },
  "middleWindBaseSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "middle wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindBaseSetting",
    "permission": "rw",
    "type": 5,
    "value": 3930,
  },
  "middleWindCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "middle wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3928,
  },
  "middleWindDirectionSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "middle wind ceiling - TRUE degrees - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindDirectionSetting",
    "permission": "rw",
    "type": 5,
    "value": 3934,
  },
  "middleWindGustEnabledSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "middle wind gust enabled - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindGustEnabledSetting",
    "permission": "rw",
    "type": 5,
    "value": 3938,
  },
  "middleWindSpeedSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "middle wind ceiling - kt - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindSpeedSetting",
    "permission": "rw",
    "type": 5,
    "value": 3932,
  },
  "middleWindTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "middle wind turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "middleWindTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3936,
  },
  "pressureDriftSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 16)",
    "description": "pressure drift - mb - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "pressureDriftSetting",
    "permission": "rw",
    "type": 2,
    "value": 3914,
  },
  "pressureQNHSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 16)",
    "description": "pressure QNH - mb - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "pressureQNHSetting",
    "permission": "rw",
    "type": 5,
    "value": 3912,
  },
  "stormCloudBaseSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "storm cloud layer base - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudBaseSetting",
    "permission": "rw",
    "type": 5,
    "value": 3886,
  },
  "stormCloudCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "storm cloud layer ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3884,
  },
  "stormCloudCoverageSetting": Offset {
    "category": "weather",
    "convert": "Math.round(({VAL} + 1) / 8192)",
    "description": "storm cloud layer coverage - oktas - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudCoverageSetting",
    "permission": "rw",
    "type": 5,
    "value": 3888,
  },
  "stormCloudIcingSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "storm cloud icing - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudIcingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3976,
  },
  "stormCloudTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "storm cloud turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3978,
  },
  "stormCloudTypeSetting": Offset {
    "category": "weather",
    "convert": "cloudType",
    "description": "storm cloud type - setting",
    "length": undefined,
    "mapping": true,
    "name": "stormCloudTypeSetting",
    "permission": "rw",
    "type": 5,
    "value": 3974,
  },
  "stormCloudVariationSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "storm cloud layer variation - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "stormCloudVariationSetting",
    "permission": "rw",
    "type": 5,
    "value": 3890,
  },
  "surfaceTemperatureLevelSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "surface temperature level - ground elevation  - setting- ft",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceTemperatureLevelSetting",
    "permission": "rw",
    "type": 5,
    "value": 3904,
  },
  "surfaceTemperatureSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "surface temperature - degrees C - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceTemperatureSetting",
    "permission": "rw",
    "type": 2,
    "value": 3906,
  },
  "surfaceWindCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "surface wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3952,
  },
  "surfaceWindDirectionSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "surface wind direction - MAG degrees - settings",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindSpeedSetting",
    "permission": "rw",
    "type": 5,
    "value": 3956,
  },
  "surfaceWindGustEnabledSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "surface wind gust enabled - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindGustEnabledSetting",
    "permission": "rw",
    "type": 5,
    "value": 3960,
  },
  "surfaceWindSpeedSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "surface wind speed - kt - settings",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindSpeedSetting",
    "permission": "rw",
    "type": 5,
    "value": 3954,
  },
  "surfaceWindTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "surface wind turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "surfaceWindTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3958,
  },
  "temperatureDayNightVariationSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "temperature day night variation - degrees C - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "temperatureDayNightVariationSetting",
    "permission": "rw",
    "type": 2,
    "value": 3910,
  },
  "temperatureDriftSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "temperature drift - degrees C - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "temperatureDriftSetting",
    "permission": "rw",
    "type": 2,
    "value": 3908,
  },
  "upperCloudBaseSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper cloud layer base - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudBaseSetting",
    "permission": "rw",
    "type": 5,
    "value": 3870,
  },
  "upperCloudCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper cloud layer ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3868,
  },
  "upperCloudCoverageSetting": Offset {
    "category": "weather",
    "convert": "Math.round(({VAL} + 1) / 8192)",
    "description": "upper cloud layer coverage - oktas - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudCoverageSetting",
    "permission": "rw",
    "type": 5,
    "value": 3872,
  },
  "upperCloudIcingSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "upper cloud icing - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudIcingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3964,
  },
  "upperCloudTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "upper cloud turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3966,
  },
  "upperCloudTypeSetting": Offset {
    "category": "weather",
    "convert": "cloudType",
    "description": "upper cloud type - setting",
    "length": undefined,
    "mapping": true,
    "name": "upperCloudTypeSetting",
    "permission": "rw",
    "type": 5,
    "value": 3962,
  },
  "upperCloudVariationSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper cloud layer variation - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperCloudVariationSetting",
    "permission": "rw",
    "type": 5,
    "value": 3874,
  },
  "upperTemperatureLevelSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper temperature level - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperTemperatureLevelSetting",
    "permission": "rw",
    "type": 5,
    "value": 3892,
  },
  "upperTemperatureSetting": Offset {
    "category": "weather",
    "convert": "Math.round({VAL} / 256)",
    "description": "upper temperature - degrees C - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperTemperatureSetting",
    "permission": "rw",
    "type": 2,
    "value": 3894,
  },
  "upperWindBaseSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindBaseSetting",
    "permission": "rw",
    "type": 5,
    "value": 3918,
  },
  "upperWindCeilingSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 3.28084).toFixed(2)",
    "description": "upper wind ceiling - ft - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindCeilingSetting",
    "permission": "rw",
    "type": 5,
    "value": 3916,
  },
  "upperWindDirectionSetting": Offset {
    "category": "weather",
    "convert": "+({VAL} * 360 / 65536).toFixed(2)",
    "description": "upper wind ceiling - TRUE degrees - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindDirectionSetting",
    "permission": "rw",
    "type": 5,
    "value": 3922,
  },
  "upperWindGustEnabledSetting": Offset {
    "category": "weather",
    "convert": "!!{VAL}",
    "description": "upper wind gust enabled - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindGustEnabledSetting",
    "permission": "rw",
    "type": 5,
    "value": 3926,
  },
  "upperWindSpeedSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "upper wind ceiling - kt - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindSpeedSetting",
    "permission": "rw",
    "type": 5,
    "value": 3920,
  },
  "upperWindTurbulenceSetting": Offset {
    "category": "weather",
    "convert": undefined,
    "description": "upper wind turbulence - 0-255 - setting",
    "length": undefined,
    "mapping": undefined,
    "name": "upperWindTurbulenceSetting",
    "permission": "rw",
    "type": 5,
    "value": 3924,
  },
  "visibilitySetting": Offset {
    "category": "weather",
    "convert": "{VAL} / 100",
    "description": "visibility setting - sm",
    "length": undefined,
    "mapping": undefined,
    "name": "visibilitySetting",
    "permission": "rw",
    "type": 5,
    "value": 3980,
  },
}
`;
